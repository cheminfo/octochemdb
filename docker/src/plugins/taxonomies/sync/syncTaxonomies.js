import { readFileSync } from 'fs';

import { fileCollectionFromZip } from 'filelist-utils';
import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import getLastFileSync from '../../../sync/http/utils/getLastFileSync.js';
import debugLibrary from '../../../utils/Debug.js';

import { getTaxonomiesNodes } from './utils/getTaxonomiesNodes.js';
import { parseTaxonomies } from './utils/parseTaxonomies.js';

const debug = debugLibrary('syncTaxonomies');
/**
 * sync NCBI taxonomies
 * @param {*} connection connection to mongo
 */
export async function sync(connection) {
  let options = {
    collectionSource: process.env.TAXONOMY_SOURCE,
    destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/taxonomies/full`,
    collectionName: 'taxonomies',
    filenameNew: 'taxonomies',
    extensionNew: 'zip',
  };
  try {
    let sources;
    let lastFile;
    if (process.env.NODE_ENV === 'test') {
      lastFile = `${process.env.TAXONOMY_SOURCE_TEST}`;
      sources = [process.env.TAXONOMY_SOURCE_TEST];
    } else {
      lastFile = await getLastFileSync(options);
      sources = [lastFile.replace(`${process.env.ORIGINAL_DATA_PATH}`, '')];
    }
    const progress = await connection.getProgress(options.collectionName);
    let isTimeToUpdate = false;
    if (
      progress.dateEnd !== 0 &&
      Date.now() - progress.dateEnd >
        Number(process.env.TAXONOMY_UPDATE_INTERVAL) * 24 * 60 * 60 * 1000 &&
      md5(JSON.stringify(sources)) !== progress.sources
    ) {
      progress.dateStart = Date.now();
      await connection.setProgress(progress);
      isTimeToUpdate = true;
    }
    const lastDocumentImported = await getLastDocumentImported(
      connection,
      progress,
      options.collectionName,
    );

    let counter = 0;
    let imported = 0;
    let start = Date.now();

    if (
      lastDocumentImported === null ||
      ((md5(JSON.stringify(sources)) !== progress.sources ||
        progress.state !== 'updated') &&
        isTimeToUpdate) ||
      process.env.NODE_ENV === 'test'
    ) {
      const collection = await connection.getCollection(options.collectionName);

      const logs = await connection.getImportationLog({
        collectionName: options.collectionName,
        sources,
        startSequenceID: progress.seq,
      });

      const fileList = (
        await fileCollectionFromZip(readFileSync(lastFile))
      ).filter((file) => file.name === 'rankedlineage.dmp');
      const arrayBuffer = await fileList.files[0].arrayBuffer();
      progress.state = 'updating';
      await connection.setProgress(progress);
      const temporaryCollection = await connection.getCollection(
        'taxonomies_tmp',
      );
      const fileListNodes = (
        await fileCollectionFromZip(readFileSync(lastFile))
      ).filter((file) => file.name === 'nodes.dmp');

      const arrayBufferNodes = await fileListNodes.files[0].arrayBuffer();
      debug('Get Nodes Taxonomies');
      let nodes = getTaxonomiesNodes(arrayBufferNodes);
      debug('start parsing taxonomies');
      for (const entry of parseTaxonomies(arrayBuffer, nodes, connection)) {
        counter++;
        if (process.env.TEST === 'true' && counter > 20) break;
        if (Date.now() - start > Number(process.env.DEBUG_THROTTLING)) {
          debug(`Processing: counter: ${counter} - imported: ${imported}`);
          start = Date.now();
        }
        entry._seq = ++progress.seq;
        await temporaryCollection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );

        imported++;
      }
      await temporaryCollection.rename(options.collectionName, {
        dropTarget: true,
      });

      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      progress.sources = md5(JSON.stringify(sources));
      progress.dateEnd = Date.now();
      progress.state = 'updated';
      await connection.setProgress(progress);
      await collection.createIndex({ 'data.phylum': 1 });
      await collection.createIndex({ 'data.class': 1 });
      await collection.createIndex({ 'data.order': 1 });
      await collection.createIndex({ 'data.family': 1 });
      await collection.createIndex({ 'data.genus': 1 });
      await collection.createIndex({ 'data.species': 1 });
      await collection.createIndex({ 'data.organism': 1 });
      await collection.createIndex({ _seq: 1 });

      debug(`${imported} taxonomies processed`);
    } else {
      debug(`file already processed`);
    }
  } catch (e) {
    if (connection) {
      debug(e.message, {
        collection: 'taxonomies',
        connection,
        stack: e.stack,
      });
    }
  }
}
