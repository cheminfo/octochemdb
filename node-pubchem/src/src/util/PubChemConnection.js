'use strict';

const delay = require('delay');
const { MongoClient } = require('mongodb');
const debug = require('debug')('PubChemConnection');

function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function getProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.updateOne({ _id: progress._id }, { $set: progress });
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    debug('Starting new database construction.');
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  } else {
    if (progress.state === 'update') {
      debug('First importation has been completed. Should only update.');
    } else {
      debug(`Continuing first importation from ${progress.seq}.`);
    }
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};

module.exports = PubChemConnection;
