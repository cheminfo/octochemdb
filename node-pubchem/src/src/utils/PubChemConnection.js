import Debug from 'debug';
import delay from 'delay';
import { MongoClient } from 'mongodb';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
