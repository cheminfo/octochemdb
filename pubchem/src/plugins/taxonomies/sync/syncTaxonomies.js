import { readFileSync } from 'fs';
import md5 from 'md5';
import { fileListFromZip } from 'filelist-from';
import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import Debug from '../../../utils/Debug.js';
import getLastFileSync from '../../../sync/http/utils/getLastFileSync.js';
import { parseTaxonomies } from './utils/parseTaxonomies.js';

const debug = Debug('syncTaxonomies');

export async function sync(connection) {
  let options = {
    collectionSource: process.env.TAXONOMY_SOURCE,
    destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/taxonomies/full`,
    collectionName: 'taxonomies',
    filenameNew: 'taxonomies',
    extensionNew: 'zip',
  };
  try {
    const lastFile = await getLastFileSync(options);
    const sources = [lastFile.replace(process.env.ORIGINAL_DATA_PATH, '')];
    const progress = await connection.getProgress(options.collectionName);
    const collection = await connection.getCollection(options.collectionName);
    const logs = await connection.geImportationtLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });
    const lastDocumentImported = await getLastDocumentImported(
      connection,
      progress,
      options.collectionName,
    );
    let firstID;
    if (lastDocumentImported !== null) {
      firstID = lastDocumentImported._id;
    }

    const fileList = (await fileListFromZip(readFileSync(lastFile))).filter(
      (file) => file.name === 'rankedlineage.dmp',
    );
    const arrayBuffer = await fileList[0].arrayBuffer();

    // we reparse all the file and skip if required
    let skipping = firstID !== undefined;
    let counter = 0;
    let imported = 0;
    let start = Date.now();
    if (
      lastDocumentImported === null ||
      md5(JSON.stringify(sources)) !== progress.sources ||
      progress.state !== 'updated'
    ) {
      for (const entry of parseTaxonomies(arrayBuffer, connection)) {
        counter++;
        if (process.env.TEST === 'true' && counter > 20) break;
        if (
          Date.now() - start >
          Number(process.env.DEBUG_THROTTLING || 10000)
        ) {
          debug(`Processing: counter: ${counter} - imported: ${imported}`);
          start = Date.now();
        }
        if (skipping) {
          if (firstID === entry._id) {
            skipping = false;
            debug(`Skipping taxonomies till:${firstID}`);
          }
          continue;
        }
        progress.state = 'updating';
        entry._seq = ++progress.seq;
        await collection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );

        await connection.setProgress(progress);
        imported++;
      }
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      progress.sources = md5(JSON.stringify(sources));
      progress.date = new Date();
      progress.state = 'updated';
      await connection.setProgress(progress);
      debug(`${imported} taxonomies processed`);
    } else {
      debug(`file already processed`);
    }

    // we remove all the entries that are not imported by the last file
    const result = await collection.deleteMany({
      _seq: { $lte: logs.startSequenceID },
    });
    debug(`Deleting entries with wrong source: ${result.deletedCount}`);
  } catch (e) {
    const optionsDebug = { collection: 'taxonomies', connection };
    debug(e, optionsDebug);
  }
}
