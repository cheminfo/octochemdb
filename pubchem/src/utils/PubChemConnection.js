import delay from 'delay';
import { MongoClient } from 'mongodb';

import Debug from './Debug.js';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getLogs = async function getCollection(
  collectionName,
) {
  const logsCollection = await this.getLogsCollection();
  const _id = `${collectionName}`;
  let logs = await logsCollection.find({ _id }).next();
  if (logs === null) {
    let sources = [];
    logs = {
      _id,
      sources,
      sourcesHash: md5(JSON.stringify(sources)),
      dataStart: 0,
      dataEnd: 0,
      startSequenceID: 123,
      endSeqneceID: 123,
      status: 'updating',
    };
    await logsCollection.insertOne(logs);
  }
  return logs;
};
PubChemConnection.prototype.setLogs = async function setLogs(logs) {
  const collection = await this.getLogsCollection();
  await collection.replaceOne({ _id: logs._id }, logs);
};
PubChemConnection.prototype.getLogsCollection = async function getCollection() {
  return (await this.getDatabase()).collection('logs');
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
