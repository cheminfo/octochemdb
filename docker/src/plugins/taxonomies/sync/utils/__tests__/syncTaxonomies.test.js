import { PubChemConnection } from '../../../../../utils/PubChemConnection.js';
import { sync } from '../../syncTaxonomies';

test('syncTaxonomies', async () => {
  const connection = new PubChemConnection();
  await sync(connection);
  const collection = await connection.getCollection('taxonomies');
  const collectionEntry = await collection.find({ _id: 2841640 }).limit(1);
  const result = await collectionEntry.next();
  expect(result).toStrictEqual({
    _id: 2841640,
    _seq: 21,
    data: {
      superkingdom: 'Archaea',
      phylum: 'Candidatus Borrarchaeota',
      class: 'Candidatus Borrarchaeia',
      order: 'Candidatus Borrarchaeales',
      family: 'Candidatus Borrarchaeaceae',
      genus: 'Candidatus Borrarchaeum',
    },
  });
  if (connection) {
    await connection.close();
  }
});
