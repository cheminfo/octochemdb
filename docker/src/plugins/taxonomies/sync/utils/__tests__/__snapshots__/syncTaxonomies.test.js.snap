// Vitest Snapshot v1

exports[`syncTaxonomies 1`] = `
{
  "_id": 2798939,
  "data": {
    "class": "Candidatus Borrarchaeia",
    "family": "Candidatus Borrarchaeaceae",
    "genus": "Candidatus Borrarchaeum",
    "order": "Candidatus Borrarchaeales",
    "phylum": "Candidatus Borrarchaeota",
    "superkingdom": "Archaea",
  },
}
`;
